### PR TITLE
Split CashPaymentFlow into a two-part flow

### PR DESCRIPTION
**Pull request by rnicoll**
Split `CashPaymentFlow` into a two-part flow so that the counterparty can confirm they're expecting the payment. This also provides hooks for inlining the transaction key negotiation.

### CODE REVIEW

**Comment by mikehearn** [comment]
Why does it need to be split in two? Can't the other side just throw an exception once a proposal is received, if it wants to reject?

**Comment by gendal** [comment]
Doesn't this completely change the semantics of payments?   If I'm understanding this correctly, the recipient of a payment now has the opportunity to hold it up?  i.e. imagine I have to pay you $1m by midnight or I'm in default and bad things will happen.  Previously, I could create the payment, sign it, get it notarised (with timestamp) and there were no games you could play to make it look like I'd failed in my duties.  Now... you just have to oh-so-accidentally take too long to respond to the PaymentProposal and now my good-faith attempt to pay you has been denied... and I don't have any independent timestamped proof that I even tried.  

Plus we've just introduced an additional round-trip...

Maybe the pros outweigh the cons (eg maybe the scenario I outline above isn't one we care about) but it feels like this PR should include some additional justification for the change.  Or the option to skip the proposal step if the sender cares about these things?

**Comment by mikehearn** [comment]
I guess that's a separate issue - being able to "force pay" someone who doesn't want to get paid seems like a feature that would need its own unit tests and observations in the code to ensure everyone who works on it realises that's a requirement (which hasn't been a concern in the past, maybe we got that feature for free because our flow was so simple).

I suspect that mistakes that cause bogus payments or repeated payments are more frequent than the sort of adversarial setup you describe there. So surely the default should be to confirm with the recipient that they're expecting a payment. In fact this sort of thing would be a benefit over current banking payment protocols that don't seem to support this at all, right? Hence a significant source of waste in UK local government being invoices that are paid more than once :)

My bigger concern here is the splitting of a flow in two. An ideal flow represents the entire life cycle of a business process, as far as the tech lets us push it e.g. without long term blocking (at the moment). We put quite some work into error propagation exactly to allow a recipient to abort flows due to exceptional conditions, so ...

**Comment by gendal** [comment]
is it actually split into two flows?  i only flicked through the code but it looked like a single flow...  and I interpreted the "split into two" as being "split into two steps".    In which case, this is fine I guess.

**Comment by rnicoll** [comment]
It's the first half of a change where transaction identity negotiation is inlined into the payment flow. It does change semantics, though, yes, but to something we've been repeatedly told is more appropriate (although I do need to actually figure how to implement the hooks for nodes to approve/reject payments, so this is useful).

And technically it splits it process into a flow that runs on each side of the transaction, if that's clearer?

**Comment by gendal** [comment]
OK... I'll step out of this one now.

File: finance/src/main/kotlin/net/corda/flows/CashPaymentFlow.kt [hunk 1]
```diff
@@ -4,62 +4,88 @@ import co.paralleluniverse.fibers.Suspendable
...
-        val anonymous: Boolean,
-        progressTracker: ProgressTracker,
-        val issuerConstraint: Set<Party> = emptySet()) : AbstractCashFlow<AbstractCashFlow.Result>(progressTracker) {
-    /** A straightforward constructor that constructs spends using cash states of any issuer. */
-    constructor(amount: Amount<Currency>, recipient: Party) : this(amount, recipient, true, tracker())
-    /** A straightforward constructor that constructs spends using cash states of any issuer. */
-    constructor(amount: Amount<Currency>, recipient: Party, anonymous: Boolean) : this(amount, recipient, anonymous, tracker())
+object CashPaymentFlow {
+    /**
+     * Initiates a flow that sends cash to a otherSide.
```
Comment (mikehearn) [inline finance/src/main/kotlin/net/corda/flows/CashPaymentFlow.kt]: to _the_ otherSide

File: finance/src/main/kotlin/net/corda/flows/CashPaymentFlow.kt [hunk 2]
```diff
@@ -4,62 +4,88 @@ import co.paralleluniverse.fibers.Suspendable
...
+            return Result(tx, anonymousRecipient)
+        }
     }
+
+    @InitiatedBy(CashPaymentFlow.Initiate::class)
+    class Receive(val otherSide: Party) : FlowLogic<Unit>() {
+        @Suspendable
+        override fun call() {
+            val proposal = receive<PaymentProposal>(otherSide)
+            // TODO: Provide hooks for KYC/AML verification logic, don't just accept everything
```
Comment (mikehearn) [inline finance/src/main/kotlin/net/corda/flows/CashPaymentFlow.kt]: Let's drop the discussions about KYC/AML. That's not really what we're getting at here, it's a separate issue. Not accepting money you aren't expecting is more about avoiding mistakes and keeping your accounting clean (no money on the books that shouldn't be there).

**Comment by rnicoll** [comment]
This seems to be stuck because the classpath scanner won't find the finance flows as they're not a full CorDapp. Shelving for now, to re-address later on.

**Comment by mikehearn** [comment]
What's the current status of this PR? Should we close it?

**Comment by rnicoll** [comment]
Was talking to Matthew about this, earlier, and it's probably taken more time than it's worth already. Closing until at least post-1.0.

**Pull request closed without merging**